server/parsec/components/memory/user.py

```diff
@@ -593,3 +597,74 @@ async def test_dump_current_users(
             )
 
         return items
+
+    async def list_users(
+        self, organization_id: OrganizationID
+    ) -> list[UserInfo] | UserListUsersBadOutcome:
+        try:
+            org = self._data.organizations[organization_id]
+        except KeyError:
+            return UserListUsersBadOutcome.ORGANIZATION_NOT_FOUND
+
+        users = []
+        for user in org.users.values():
+            users.append(
+                UserInfo(
+                    user_id=user.cooked.user_id,
+                    human_handle=user.cooked.human_handle,
+                    frozen=user.is_frozen,
+                )
+            )
+
+        return users
+
+    async def freeze_user(
+        self,
+        organization_id: OrganizationID,
+        user_id: UserID | None,
+        user_email: str | None,
+        frozen: bool,
+    ) -> UserInfo | UserFreezeUserBadOutcome:
+        try:
+            org = self._data.organizations[organization_id]
+        except KeyError:
+            return UserFreezeUserBadOutcome.ORGANIZATION_NOT_FOUND
+
+        match (user_id, user_email):
+            case (None, None):
+                return UserFreezeUserBadOutcome.NO_USER_ID_NOR_EMAIL
+            case (UserID() as user_id, None):
+                try:
+                    user = org.users[user_id]
+                except KeyError:
+                    return UserFreezeUserBadOutcome.USER_NOT_FOUND
+            case (None, str() as user_email):
+                for user in org.users.values():
+                    if user.cooked.human_handle.email == user_email:
+                        break
+                else:
+                    return UserFreezeUserBadOutcome.USER_NOT_FOUND
+            case _:
+                return UserFreezeUserBadOutcome.BOTH_USER_ID_AND_EMAIL
```

Comment:
No exhaustive match here?

Reply:
Exhaustive matching on tuple is not supported (see https://github.com/python/mypy/issues/16650)

Reply:
That's sad. Maybe you can still write it with:
```python
case _: 
    assert False  # Can't use assert_never here due to https://github.com/python/mypy/issues/16650
```